### ISSUE TITLE
Why can attribute accessors only use enclosing type's type variables and not have their own generic type??

### ISSUE THREAD

**Issue by san2488** [immutables/immutables#952]
Why is [this check](https://github.com/immutables/immutables/blob/3daf8b671e32d4cecfee6ca25662070e7581d1b4/value-processor/src/org/immutables/value/processor/meta/AccessorAttributesCollector.java#L244) required to restrict accessor methods from defining their own generic types?

I'd like to do something like this:
```java
@Value.Immutable
interface MyAmazingClass<T> {

<R extends T> getMostAmazingField();

T getMoreAmazingField();
}
```

But seems like that is not possible with Immutables. What is the downside of supporting it?

**Comment by elucash** [comment]
Thank you for the question. Here's how I can try to explain it: the field can only store something which type is known at a lifetime of the instance. The generic type R introduced for `getMostAmazingField` is being evaluated only at a call-site of this method, R can be substituted for different type extends T while still storing the same instance of T (of some subtype of it). So if T is substituted with type A, and you pass B extends A and then use B as R when calling `getMostAmazingField`, there's unchecked cast would happen and it is not safe (stored A instance can be of type C extends A, assuming R as B during the call will lead to class cast on retrieval.
basically, what you probably would need to do is something like this.
```
@Value.Immutable
interface MyAmazingClass<T> {

T getMostAmazingField();

@SuppressWarning("unchecked") // it's just unsafe
<R extends T> R getUncheckedMostAmazingField() {
   return (R) getMostAmazingField();
}
}
```

Immutables would not generate code known to be unsafe. Basically, the workaround would stick to just T, cast to R manually, at least problems would not be hidded in generated code.

**Comment by san2488** [comment]
That makes sense. Thanks!
